File: reverseInteger/RevInt.java
/**
 * https://leetcode.com/problems/reverse-integer/ : Given a 32-bit signed integer,
 * reverse digits of an integer. Assume we are dealing with an environment which
 * could only store integers within the 32-bit signed integer range: [−2^31,
 * 2^31 − 1]. For the purpose of this problem, assume that your function returns
 * 0 when the reversed integer overflows.
 */
public class RevInt {
    public int reverse(int x) {
        long copy = x, rev = 0;

        // reverse the number digit-by-digit
        while (copy != 0) {
            rev = rev * 10 + copy % 10;
            copy = copy / 10;
        }

        // overflow case
        if (rev > Integer.MAX_VALUE || rev < Integer.MIN_VALUE)
            return 0;
        else
            return (int) (rev); // convert long to int and return

    }
}

File: src/reverseInteger/revInt.ts
// reverse a 32-bit signed integer; return 0 if reversal goes outside range
function reverse(x: number): number {
  const digits: string = String(x).substring(x < 0 ? 1 : 0);
  const rev: number = parseInt(digits.split("").reverse().join(""), 10);

  const is32BitInt = (n: number): boolean => n >= -(2 ** 31) && n <= 2 ** 31 - 1;
  const fixSign = (x: number, rev: number): number => (x < 0 ? -rev : rev);

  return is32BitInt(rev) ? fixSign(x, rev) : 0;
}

function test(x: number): void {
  console.log(`Input: ${x}\nOutput: ${reverse(x)}\n`);
}

test(123);
test(-123);
test(120);
test(0);
test(1534236469);
